### PR TITLE
[codex] Fix stale app cleanup for image-less GVCs

### LINKED ISSUE #309: CI quota exhaustion: cleanup-stale-apps skips image-less orphan GVCs + no concurrency control on shared test org
## Summary

CI runs intermittently fail with `Quota <gvcs> has been maxed out (max=20)` because the test org `shakacode-heroku-to-control-plane-ci` accumulates orphan `dummy-test-*` GVCs that `cleanup-stale-apps` cannot detect. Compounded by concurrent CI runs (8+ parallel jobs observed) sharing a single 20-GVC quota with no concurrency control. See failed run [77100924233](https://github.com/shakacode/control-plane-flow/actions/runs/26204353662) — all 19 failures share the same 429 root cause.

## Algorithm bug in `cleanup-stale-apps`

`lib/command/cleanup_stale_apps.rb:47-57`:

```ruby
gvcs = cp.gvc_query(config.app)["items"]
gvcs.each do |gvc|
  app_name = gvc["name"]

  images = cp.query_images(app_name)["items"].select { |item| item["name"].start_with?("#{app_name}:") }
  image = cp.latest_image_from(images, app_name: app_name, name_only: false)
  next unless image                                  # <-- bug: skips GVCs without images

  created_date = DateTime.parse(image["created"])
  diff_in_days = (now - created_date).to_i
  next unless diff_in_days >= stale_app_image_deployed_days
  ...
end
```

Staleness is determined by **latest image creation date**, with `next unless image`. Consequences:

1. **GVCs that crashed before image build are immortal.** Quota-throttled `cpflow setup-app` (which is exactly the failure mode triggering this issue) leaves a partial GVC behind with no image → cleanup permanently skips it.
2. **Test scenarios that never build images are immortal.** `spec/command/delete_spec.rb`, `spec/command/info_spec.rb`, etc. create apps via `apply-template` but never `build-image`. Their GVCs have no images, so cleanup skips them.
3. **Granularity is days, not minutes.** `stale_app_image_deployed_days` is integer days. For a 25-minute CI suite that burns quota every run, day-granularity cleanup is useless as a same-day mitigation.

Spotted in production: at the time of writing, the test org has `dummy-test-valid-pre-deletion-hook-52d5-e3d7` (26 min old, no images, suffix unmatched by any other GVC = run completed). `cleanup-stale-apps` would never touch it.

## Compounding: concurrent runs starve each other

`gh run list --status in_progress` showed 8 simultaneous RSpec jobs against the same Control Plane org. Each run creates ~5 GVCs across its ~25 min lifecycle: 8 × 5 = 40 needed vs. 20 ceiling. There is no `concurrency:` group on the RSpec workflow to serialize them.

## Why `after(:suite)` cleanup isn't enough

`spec/support/dummy_app_setup.rb` only deletes apps tracked in the in-memory `apps_to_delete` array via the `after(:suite)` hook. If the suite is killed mid-run (timeout, runner OOM, cancelled job, network blip), the cleanup never runs and every GVC created so far leaks.

## Recommended fixes (pick any combination)

1. **Fix `cleanup-stale-apps` to fall back to GVC `created` timestamp when no images exist.** Replace `next unless image` with: if no image, use `DateTime.parse(gvc["created"])` instead. This single change makes the existing tool effective against the orphan failure mode.

2. **Add `--max-age-minutes` (or `--max-age-hours`) option** for fine-grained cleanup. Day granularity is wrong for CI orphans that need cleanup within an hour.

3. **Add a pre-suite sweep in `DummyAppSetup.setup`** (`spec/support/dummy_app_setup.rb`) that deletes all `dummy-test-*` GVCs older than ~60 min before the suite runs. Self-healing — orphans from any prior crashed run get reaped automatically. Skipped locally via `SKIP_CLEANUP=true` like the existing teardown.

4. **Add a `concurrency:` group on the RSpec workflow** (`.github/workflows/rspec*.yml`) keyed on `cpln-shared-org` so only one job hits the shared org at a time. Trade-off: PRs queue.

5. **Request a GVC quota increase** from Control Plane admins (e.g., 50 or 100). Doesn't fix the orphan-accumulation bug but raises the ceiling.

6. **Long-term**: per-PR ephemeral test org, removing the shared-quota constraint entirely.

## Related work

- PR #302 / issue #295 adds `--mode=stop` to `cleanup-stale-apps`. Useful for review apps, but doesn't change the staleness-detection algorithm and therefore doesn't help with the CI orphan accumulation described here.

## Repro signal for the fix

After fix #1 lands, running `cpflow cleanup-stale-apps -a <prefix> --yes` against the CI org should remove orphan dummy-test GVCs that have no images. Verify with `cpln gvc get --org shakacode-heroku-to-control-plane-ci` before/after.

### PR DESCRIPTION
**Pull request by justin808**
Fixes #309.

This updates `cleanup-stale-apps` to treat image-less GVCs as stale based on the GVC `created` timestamp, so partial CI apps are no longer skipped forever. It also serializes shared RSpec workflow jobs by `vars.CPLN_ORG` to reduce concurrent CI pressure on the same Control Plane GVC quota. Added regression coverage for the cleanup fallback and workflow concurrency, and updated README/generated command docs.

Validation: `CPLN_ORG=test-org bundle exec rspec spec/command/cleanup_stale_apps_spec.rb:29 spec/github_workflows_spec.rb`, `bundle exec rubocop lib/command/cleanup_stale_apps.rb spec/command/cleanup_stale_apps_spec.rb spec/github_workflows_spec.rb`, `script/check_command_docs`, and `git diff --check origin/main...`.

---

> [!NOTE]
> **Low Risk**
> Low risk: test-only changes that add coverage for edge cases without modifying production behavior.
> 
> **Overview**
> Adds specs for `Command::CleanupStaleApps#stale_apps` to ensure apps with no images fall back to the GVC `created` timestamp, and that GVCs with a missing `created` value are skipped instead of raising.
> 
> Also simplifies the `Config` test double setup by removing the unused `options: { yes: true }` stub in this spec.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 3582d0fd2f1e1621a7f2156c2319f372b306e687. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: .github/workflows/rspec-shared.yml [hunk 1]
```diff
@@ -19,6 +19,9 @@ on:
 jobs:
   rspec:
     runs-on: ${{ inputs.os_version }}
+    concurrency:
+      group: cpln-shared-org-${{ vars.CPLN_ORG }}
```
[the remark this answers is not in the document]
Comment (justin808) [reply .github/workflows/rspec-shared.yml]: Fixed in 111b999 — added `|| github.run_id` fallback so forks without `CPLN_ORG` configured no longer collapse into a shared global queue. Also updated `spec/github_workflows_spec.rb` to match.

File: spec/command/cleanup_stale_apps_spec.rb [hunk 2]
```diff
@@ -6,6 +6,32 @@
...
+      expect(command.send(:stale_apps)).to eq([
+                                                {
+                                                  name: gvc.fetch("name"),
+                                                  date: DateTime.parse(gvc.fetch("created"))
+                                                }
+                                              ])
+    end
+  end
+
   context "when 'stale_app_image_deployed_days' is not defined" do
```
[the remark this answers is not in the document]
Comment (justin808) [reply spec/command/cleanup_stale_apps_spec.rb:43]: Fixed in 111b999 — updated the `lists correct stale apps` integration test to reflect the new behaviour. `app4` (image-less GVC, old GVC date) is now expected to be classified as stale via the fallback path, so the confirmation matcher now checks for `3 apps` and `app4` is asserted to appear in stderr.

**Comment by justin808** [comment]
## Review addressed in 111b999

**1. Concurrency group fallback ([thread](https://github.com/shakacode/control-plane-flow/pull/310#discussion_r3293576689))**
`.github/workflows/rspec-shared.yml`: added `vars.CPLN_ORG || github.run_id` fallback so forks without the variable configured don't collapse into a single global queue. Workflow spec updated to match.

**2. `lists correct stale apps` integration test ([thread](https://github.com/shakacode/control-plane-flow/pull/310#discussion_r3293577040))**
`spec/command/cleanup_stale_apps_spec.rb`: updated expectations to reflect the new fallback behaviour. `app4` (image-less GVC with an old GVC date) is now classified as stale via the fallback path, so the test asserts 3 stale apps and includes `app4` in expected stderr output. This actually strengthens the integration test by exercising the new code path end-to-end.

Note: I couldn't run `:slow` integration tests locally (no `CPLN_ORG`), but the unit test at line 25 and `spec/github_workflows_spec.rb` both pass.

File: spec/command/cleanup_stale_apps_spec.rb [hunk 3]
```diff
@@ -6,6 +6,32 @@
 describe Command::CleanupStaleApps do
   let!(:app_prefix) { dummy_test_app_prefix("stale-app") }
 
+  describe "#stale_apps" do
+    let(:config) { instance_double(Config, app: "dummy-test", options: { yes: true }) }
```
[the remark this answers is not in the document]
Comment (justin808) [reply spec/command/cleanup_stale_apps_spec.rb]: Fixed in 3582d0f — dropped `options: { yes: true }` from the `config` double since `stale_apps` doesn't read it.

File: spec/command/cleanup_stale_apps_spec.rb [hunk 4]
```diff
@@ -6,6 +6,32 @@
...
+      allow(config).to receive(:[]).with(:stale_app_image_deployed_days).and_return(5)
+      allow(command).to receive(:cp).and_return(cp)
+      allow(cp).to receive(:gvc_query).with("dummy-test").and_return({ "items" => [gvc] })
+      allow(cp).to receive(:query_images).with(gvc.fetch("name")).and_return({ "items" => [] })
+      allow(cp).to receive(:latest_image_from)
+        .with([], app_name: gvc.fetch("name"), name_only: false)
+        .and_return(nil)
+    end
+
+    it "uses the GVC creation date when the app has no images" do
```
[the remark this answers is not in the document]
Comment (justin808) [reply spec/command/cleanup_stale_apps_spec.rb:25]: Fixed in 3582d0f — added a `context "when the GVC has no creation date"` covering the `next unless created_at` guard so the nil-`created` branch is exercised.

**Comment by justin808** [comment]
## Review nits addressed in 3582d0f

**1. Unused `options` stub ([thread](https://github.com/shakacode/control-plane-flow/pull/310#discussion_r3293598732))**
`spec/command/cleanup_stale_apps_spec.rb:10`: dropped `options: { yes: true }` from the `Config` double — `stale_apps` only calls `config.app` and `config[:stale_app_image_deployed_days]`.

**2. Nil-`created` branch coverage ([thread](https://github.com/shakacode/control-plane-flow/pull/310#discussion_r3293598759))**
`spec/command/cleanup_stale_apps_spec.rb:34-40`: added `context "when the GVC has no creation date"` that exercises the `next unless created_at` guard and asserts the app is skipped (not raised on).

Re: the concurrency-group note about matrix dimensions — that one is intentionally coarse so all matrix legs share the single org-wide CPLN quota. Leaving as-is; happy to add a YAML comment if you'd prefer that explicit.

Verified locally:
- `bundle exec rubocop spec/command/cleanup_stale_apps_spec.rb` → no offenses
- `CPLN_ORG=dummy bundle exec rspec spec/command/cleanup_stale_apps_spec.rb:25 :35` → 2 examples, 0 failures
